src/utils.jl

```diff
@@ -313,50 +202,104 @@ function split_dataset(data::Matrix; train_ratio::Float64 = 0.9, randomize::Bool
   return train_data,test_data
 end
 
+"""
+    _ischoi(M::LPDO)
 
+Check whether a given LPDO{MPO}  
+"""
+_ischoi(M::LPDO{MPO}) = (length(inds(M.X[1],"Site")) == 2 && hastags(M.X[1],"Purifier"))
 
-function ischoi(M::LPDO)
-  return (length(inds(M.X[1],"Site")) == 2 ? true : false)
-end
+_ischoi(M::MPO) = length(inds(M[1],"Site")) == 4 
+
+
+
+"""
+    _haschoitags(L::LPDO)
+
+    _haschoitags(M::Union{MPS,MPO})
+
+Check whether the TN has input/output Choi tags
+"""
+_haschoitags(L::LPDO) = (hastags(inds(L.X[1]),"Input") && hastags(inds(L.X[1]),"Output"))
+_haschoitags(M::Union{MPS,MPO}) = _haschoitags(LPDO(M)) 
 
-function ischoi(M::MPO)
-  return (length(inds(M[1],"Site")) == 4 ? true : false)
-  #return ( length(inds(M[1])) == 5 ? true : false)
-end
 
-function makeUnitary(L::LPDO{MPS})
-  ψ = L.X
-  U = MPO(ITensor[copy(ψ[j]) for j in 1:length(ψ)])
-  prime!(U,tags="Output")
-  removetags!(U, "Input")
-  removetags!(U, "Output")
-  return U
+"""
+    _choitags(U::Union{MPS, MPO})
+
+Assign the input/output tags defined for a Choi matrix to an MPO.
+
+  σ₁ -o- σ₁′       σ₁ⁱ -o- σ₁ᴼ   
+      |                 |
+  σ₂ -o- σ₂′  ⟶    σ₂ⁱ -o- σ₂ᴼ
+      |                 |
+  σ₃ -o- σ₃′       σ₃ⁱ -o- σ₃ᴼ
+                  
+"""
+function _choitags(L::LPDO{MPO})
+  _haschoitags(L) && return L
+  U = copy(L.X)
+  U = addtags(U, "Input", plev = 0, tags = "Qubit")
```

Comment:
Better to make this `tags = "Site"` instead of `tags = "Qubit"` (in theory these functions aren't particular to the Qubit type).